e2e/stack_test.go

```diff
@@ -0,0 +1,136 @@
+package e2e_test
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/big"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/polymerdao/monomer/e2e"
+	"github.com/polymerdao/monomer/utils"
+	"github.com/sourcegraph/conc"
+	"github.com/stretchr/testify/require"
+)
+
+type logWriter struct {
+	t           *testing.T
+	programName string
+}
+
+func newLogWriter(t *testing.T, programName string) *logWriter {
+	return &logWriter{
+		t:           t,
+		programName: programName,
+	}
+}
+
+func (w *logWriter) Write(data []byte) (int, error) {
+	w.t.Log(w.programName + ": " + string(data))
+	return len(data), nil
+}
+
+func TestE2E(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping e2e tests in short mode")
+	}
+
+	const (
+		l1Port uint16 = iota + 8888
+		monomerPort
+		opNodePort
+	)
+	const l1BlockTime = time.Second // We would want 250ms instead, but then Anvil doesn't respond to RPCs (probably a thread starving in Anvil).
+	monomerAddress := newAddress(monomerPort)
+
+	_, verbose := os.LookupEnv("MONOMER_E2E_VERBOSE")
+
+	contractsRootDir, err := filepath.Abs("./optimism/packages/contracts-bedrock/")
+	require.NoError(t, err)
+
+	stack := e2e.New(newAddress(l1Port), monomerAddress, newAddress(opNodePort), contractsRootDir, l1BlockTime, &e2e.SelectiveListener{
+		OPLogWithPrefixCb: func(prefix string, r *log.Record) {
+			if prefix != "node" && !verbose {
+				return
+			}
+			r.Msg = prefix + ": " + r.Msg
+			t.Log(string(log.TerminalFormat(false).Format(r)))
+		},
+		AfterOPStartupCb: func() {
+			t.Log("started op stack")
+		},
+		BeforeOPShutdownCb: func() {
+			t.Log("shutting down op stack")
+		},
+		OnCmdStartCb: func(programName string, stdout, stderr io.Reader) {
+			t.Log("started " + programName)
+			if verbose {
+				out := newLogWriter(t, programName)
+				go func() {
+					_, err := io.Copy(out, stdout)
+					require.NoError(t, err)
+				}()
+				go func() {
+					_, err := io.Copy(out, stderr)
+					require.NoError(t, err)
+				}()
+			}
+		},
+		OnCmdStoppedCb: func(programName string, err error) {
+			if err == nil {
+				t.Log("successfully stopped " + programName)
+			} else {
+				t.Logf("stopped %s with: %v", programName, err)
+			}
+		},
+	})
+
+	var wg conc.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg.Go(func() {
+		require.NoError(t, stack.Run(ctx))
+	})
+	require.True(t, utils.AddressIsReady(monomerAddress, 10000000))
```

Comment:
Haha, yeah I should probably add a comment for this one. It's deliberate. What do you think about the following?

> To avoid flaky tests, hang until the Monomer server is ready. We rely on the `go test` timeout to ensure the tests don't hang forever (default is 10 minutes).